### ISSUE TITLE
[SITUP Core] Avoid requirement of empty constructor in Plugins

### ISSUE THREAD

**Issue by kowshikn** [opendistro-for-elasticsearch/simple-ingest-transformation-utility-pipeline#66]
There is a need for mandatory constructor (like below) in Plugins for the PipelineConfiguration to init the plugin, we can avoid this.

```
/**
     * Mandatory constructor for SITUP Component - This constructor is used by SITUP
     * runtime engine to construct an instance of {@link StringProcessor} using an instance of {@link PluginSetting} which
     * has access to pluginSetting metadata from pipeline
     * pluginSetting file.
     *
     * @param pluginSetting instance with metadata information from pipeline pluginSetting file.
     */
    public StringProcessor(final PluginSetting pluginSetting) {
        this();
    }

    public StringProcessor() {

    }
```

**Comment by yadavcbala** [comment]
Yes, this will be removed when we move to ServiceLoader pattern which is currently below the line.

**Comment by yadavcbala** [comment]
Updating the title as it is says *empty constructor*, currently there is no requirement of empty constructor - requirement is to have a constructor with PluginSetting as the only parameter.

**Comment by kowshikn** [comment]
Linking https://github.com/opendistro-for-elasticsearch/data-prepper/issues/252 regarding using DI. We originally decided to avoid DIs, but we can discuss this when we touch this story again.

**Comment by wrijeff** [comment]
No need to discuss dependency injection frameworks as part of this issue - @kowshikn was correct in that my suggestion to check out DI's wouldn't really help here.

From issue #252, part of the pain I ran into when creating a more complex plugin was being constrained by the requirement that the factory classes can only create objects by calling `constructor.newInstance(pluginSetting)` on the actual plugin itself. This makes it difficult to use dependency _inversion_ in that the basic constructor ultimately has to call `new` a bunch of times. Example from #252, (not real code):
```
public PeerForwarder(final PluginSetting pluginSetting) {
    super(pluginSetting);
    this.peerClientPool = new PeerClientPool(); 
    this.hashRing = new HashRing(); 
    // More complex logic to create necessary objects
}
```

---


One idea I'd like to discuss would be requiring plugins to define factory classes (extending off of predefined abstract factory classes likely residing in the API package) which build and return the plugins to be used in a pipeline, instead of requiring them to implement the single-arg constructor. Essentially we'd be factoring out instantiation logic into a separate layer: 

PipelineParser -> Find factory by annotation -> Call `factory.createPlugin(pluginSetting)` -> Return a fully-built plugin object

Where the factory would look something like:
```
public class PeerForwarderFactory extends AbstractProcessorFactory {
    
    @Override
    public Processor createProcessor(final PluginSetting pluginSetting) {

        // Instantiation logic

        return new PeerForwarder(peerClientPool,
                hashRing,
                peerListProvider);
    }
    
}
```
This would let us utilize dependency inversion on the actual plugin classes, such as:
```
public PeerForwarder(final PluginSetting pluginSetting,
                        final PeerClientPool peerClientPool,
                        final HashRing hashRing,
                        final PeerListProvider peerListProvider {
    super(pluginSetting);
    this.peerClientPool = peerClientPool;
    this.hashRing = hashRing;
    this.peerListProvider = peerListProvider;
}
```

Each plugin project would be required to define both the plugin class _and_ the factory class responsible for creating the actual plugin object. Data Prepper core objects (i.e. PipelineParser) would use the plugin's respective factory to instantiate the plugin and its dependencies, instead of requiring the plugin to do the heavy lifting of dependency creation.

Just a thought that occurred while creating the PeerForwarder processor and its unit tests, would be interested how previous discussions around plugins went.

**Comment by kowshikn** [comment]
After multiple discussions this story has been pushed to Q3 2021.
